### ISSUE TITLE
False positive: TOP_OF_PIPE should behave the same as NONE in the first scope

### ISSUE THREAD

**Issue by theakman2** [KhronosGroup/Vulkan-ValidationLayers#7353]
**Environment:**
 - OS: Linux 6.5.0-14-generic x86_64 x86_64 (Xubuntu 23.10)
 - GPU and driver version: NVIDIA GeForce RTX 3070 Ti, 535.146.2.0
 - SDK or header version if building from repo: 1.3.275
 - Options enabled (synchronization, best practices, etc.): 'Synchronization Preset' from vkconfig

**Describe the Issue**

SYNC-HAZARD-WRITE-AFTER-READ reported when there is no hazard. See below for exact error reported.

The error is (incorrectly) reported when I have the following pipeline barrier:

```
vke::cmdPipelineBarrier(
	drawBuf, // cmd buf
	m_colorImages[frameIndex].getImage(), // image
	vkw::ImageLayout::eUndefined, // old layout
	vkw::ImageLayout::eColorAttachmentOptimal, // new layout
	vkw::PipelineStageFlagBits2::eNone, // src stage mask (i.e. VK_PIPELINE_STAGE_2_NONE)
	vkw::AccessFlagBits2::eNone, // src access mask
	vkw::PipelineStageFlagBits2::eColorAttachmentOutputBit, // dst stage mask
	vkw::AccessFlagBits2::eColorAttachmentWriteBit // dst access mask
);
```

Changing the source stage mask from `vkw::PipelineStageFlagBits2::eNone` (`VK_PIPELINE_STAGE_2_NONE`) to `vkw::PipelineStageFlagBits2::eTopOfPipeBit` (i.e. `VK_PIPELINE_STAGE_2_TOP_OF_PIPE_BIT`) removes the validation warnings.

However the Vulkan specification (https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineStageFlagBits2.html) states:

> VK_PIPELINE_STAGE_2_TOP_OF_PIPE_BIT is ... equivalent to VK_PIPELINE_STAGE_2_NONE in the first scope.

I would therefore expect neither VK_PIPELINE_STAGE_2_TOP_OF_PIPE_BIT nor VK_PIPELINE_STAGE_2_NONE to cause validation warnings.

**Expected behavior**

The validation layers should not report a write-after-read hazard when the source scope is `VK_PIPELINE_STAGE_2_NONE` in this case.

**Valid Usage ID**

See below for exact validation message reported.

```
SYNC-HAZARD-WRITE-AFTER-READ(ERROR / SPEC): msgNum: 929810911 - Validation Error: [ SYNC-HAZARD-WRITE-AFTER-READ ] Object 0: handle = 0x5190000d7ad0, name = Queue: Graphics and present, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x376bc9df | vkQueueSubmit2():  Hazard WRITE_AFTER_READ for entry 0, VkCommandBuffer 0x522000012150[], Submitted access info (submitted_usage: SYNC_IMAGE_LAYOUT_TRANSITION, command: vkCmdPipelineBarrier2, seq_no: 1, VkImage 0x51a0000174d0[Image: Color image 0], reset_no: 50). Access info (prior_usage: SYNC_FRAGMENT_SHADER_SHADER_SAMPLED_READ, read_barriers: VK_PIPELINE_STAGE_2_TOP_OF_PIPE_BIT|VK_PIPELINE_STAGE_2_DRAW_INDIRECT_BIT|VK_PIPELINE_STAGE_2_VERTEX_SHADER_BIT|VK_PIPELINE_STAGE_2_TESSELLATION_CONTROL_SHADER_BIT|VK_PIPELINE_STAGE_2_TESSELLATION_EVALUATION_SHADER_BIT|VK_PIPELINE_STAGE_2_GEOMETRY_SHADER_BIT|VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT|VK_PIPELINE_STAGE_2_EARLY_FRAGMENT_TESTS_BIT|VK_PIPELINE_STAGE_2_LATE_FRAGMENT_TESTS_BIT|VK_PIPELINE_STAGE_2_COLOR_ATTACHMENT_OUTPUT_BIT|VK_PIPELINE_STAGE_2_COMPUTE_SHADER_BIT|VK_PIPELINE_STAGE_2_BOTTOM_OF_PIPE_BIT|VK_PIPELINE_STAGE_2_HOST_BIT|VK_PIPELINE_STAGE_2_COMMAND_PREPROCESS_BIT_NV|VK_PIPELINE_STAGE_2_CONDITIONAL_RENDERING_BIT_EXT|VK_PIPELINE_STAGE_2_TASK_SHADER_BIT_EXT|VK_PIPELINE_STAGE_2_MESH_SHADER_BIT_EXT|VK_PIPELINE_STAGE_2_RAY_TRACING_SHADER_BIT_KHR|VK_PIPELINE_STAGE_2_FRAGMENT_SHADING_RATE_ATTACHMENT_BIT_KHR|VK_PIPELINE_STAGE_2_FRAGMENT_DENSITY_PROCESS_BIT_EXT|VK_PIPELINE_STAGE_2_TRANSFORM_FEEDBACK_BIT_EXT|VK_PIPELINE_STAGE_2_ACCELERATION_STRUCTURE_BUILD_BIT_KHR|VK_PIPELINE_STAGE_2_VIDEO_DECODE_BIT_KHR|VK_PIPELINE_STAGE_2_VIDEO_ENCODE_BIT_KHR|VK_PIPELINE_STAGE_2_ACCELERATION_STRUCTURE_COPY_BIT_KHR|VK_PIPELINE_STAGE_2_OPTICAL_FLOW_BIT_NV|VK_PIPELINE_STAGE_2_MICROMAP_BUILD_BIT_EXT|VK_PIPELINE_STAGE_2_COPY_BIT|VK_PIPELINE_STAGE_2_RESOLVE_BIT|VK_PIPELINE_STAGE_2_BLIT_BIT|VK_PIPELINE_STAGE_2_CLEAR_BIT|VK_PIPELINE_STAGE_2_INDEX_INPUT_BIT|VK_PIPELINE_STAGE_2_VERTEX_ATTRIBUTE_INPUT_BIT|VK_PIPELINE_STAGE_2_SUBPASS_SHADER_BIT_HUAWEI|VK_PIPELINE_STAGE_2_INVOCATION_MASK_BIT_HUAWEI|VK_PIPELINE_STAGE_2_CLUSTER_CULLING_SHADER_BIT_HUAWEI, queue: VkQueue 0x5190000d7ad0[Queue: Graphics and present], submit: 92, batch: 1, batch_tag: 600, command: vkCmdDraw, command_buffer: VkCommandBuffer 0x522000012150[], seq_no: 8, reset_no: 48).
    Objects: 1
        [0] 0x5190000d7ad0, type: 4, name: Queue: Graphics and present
```

**Additional context**

Between the prior usage of SHADER_SAMPLED_READ in the fragment shader (as indicated in the error message above) and the subsequent pipeline barrier mentioned above, my renderer signals and waits on a timeline semaphore with stageMask=VK_PIPELINE_STAGE_2_ALL_COMMANDS_BIT (for both signal and wait operations).

According to my understanding of the specification (https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#synchronization-semaphores-waiting), this should ensure the subsequent write happens-after the prior SHADER_SAMPLED_READ read, and also ensure that any memory accesses are made available and visible.

My renderer uses dynamic rendering.

**Comment by RYDB3RG** [comment]
sounds a bit like #7129

**Comment by theakman2** [comment]
My renderer uses dynamic rendering, so that's possible.

**Comment by artem-lunarg** [comment]
@theakman2 At the moment synchronization validation does not support timeline semaphores, sorry for this. I have plans to work on support somewhere this year.

**Comment by artem-lunarg** [comment]
@theakman2 Here are results of initial investigation. In the above example the error is expected, explained below. Also there is a potential bug on the VVL side too.

Added repro cases both for binary and timeline semaphores: https://github.com/artem-lunarg/Vulkan-ValidationLayers/commit/012da6469cf5665a644007b6a75c6ea57d7f3eaa

@RYDB3RG is correct, this issue is similar to 7193 (it was not about dynamic rendering). It's a bit cursed side of a Vulkan synchronization. In Vulkan, when synchronization primitive (e.g. semaphore or pipeline barrier) specifies `pipeline stage` mask it synchronizes operations that can **run on those stages**. A special feature of a barrier command (e.g. `vkCmdPipelineBarrier`) is that it *does not run on any pipeline stage*. If semaphore specifies `ALL_COMMANDS` then the layout transition command is not in a synchronization scope of that semaphore automatically, because layout transition does not run on any of the stages described by `ALL_COMMANDS`. The regular commands (like `vkCmdDraw`) are automatically included in sync scope based on the stages where they run. A synchronization command **should specify execution dependency** using its pipeline stage parameters in order to chain with sync scopes specified by other synchronization primitives.

The semaphore in this issue specifies `ALL_COMMANDS` for a signal and wait, and the next submission batch starts with layout transition barrier that specifies `NONE` as source stage. It means that transition does not create execution dependency with semaphore and can execute earlier, thus hazarding with shader image read operation in the previous submission batch. In this scenario we expect WRITE-AFTER-READ hazard. The app can fix the issue by specifying arbitrary stage mask that has non-empty overlap with `ALL_COMMANDS` (including `ALL_COMMANDS` value itself) as the source stage of layout transition.

The problem on the VVL side is that `TOP_OF_PIPE` stage for some reason creates execution dependency with `ALL_COMMANDS`, but as you mentioned above the spec says it should behave as NONE in the first sync scope. I'm going to investigate this further. 

p.s. Sync validation does not properly support timeline semaphores yet, but in some cases it works because of luck. I provided test example both for timeline and binary semaphore and they both reproduce the issue.

**Comment by artem-lunarg** [comment]
@theakman2  After internal discussion it looks like my explanation was incorrect. ALL_COMMANDS should be fine to synchronize submission. Apologies for confusion, also removed the previous answer in order not to confuse people more. Reopening this issue.

**Comment by thevaber** [comment]
I encountered a similar issue when doing an image layout transition after a semaphore waits for ALL_COMMANDS when srcStageMask for the layout transition barrier is NONE, and #7465 does not fix it (I think the main difference is that this is a WRITE-AFTER-WRITE hazard instead of WRITE-AFTER-READ), here's a short test case which reproduces the issue (with some additional comments):

```
TEST_F(PositiveSyncVal, TestLayoutTransitionAllCommandsWaitSemaphore) {
    SetTargetApiVersion(VK_API_VERSION_1_2);
    AddRequiredExtensions(VK_KHR_SYNCHRONIZATION_2_EXTENSION_NAME);
    AddRequiredFeature(vkt::Feature::synchronization2);
    RETURN_IF_SKIP(InitSyncValFramework());
    RETURN_IF_SKIP(InitState());
    QSTestContext test(m_device);

    vkt::Image image(*m_device, 64, 64, 1, VK_FORMAT_R8_UNORM, VK_IMAGE_USAGE_TRANSFER_DST_BIT);

    VkImageMemoryBarrier2KHR image_barrier = vku::InitStructHelper();
    image_barrier.srcStageMask = VK_PIPELINE_STAGE_2_NONE;
    image_barrier.dstStageMask = VK_PIPELINE_STAGE_2_TRANSFER_BIT;
    image_barrier.srcAccessMask = VK_ACCESS_2_NONE;
    image_barrier.dstAccessMask = VK_ACCESS_2_TRANSFER_WRITE_BIT_KHR;
    image_barrier.oldLayout = VK_IMAGE_LAYOUT_UNDEFINED;
    image_barrier.newLayout = VK_IMAGE_LAYOUT_GENERAL;
    image_barrier.image = image;
    image_barrier.subresourceRange = {VK_IMAGE_ASPECT_COLOR_BIT, 0, 1, 0, 1};
    VkDependencyInfoKHR dep_info = vku::InitStructHelper();
    dep_info.imageMemoryBarrierCount = 1;
    dep_info.pImageMemoryBarriers = &image_barrier;

    test.BeginA();
    vk::CmdPipelineBarrier2KHR(test.h_cba, &dep_info);
    vk::CmdClearColorImage(test.h_cba, image, VK_IMAGE_LAYOUT_GENERAL, &m_clear_color, 1, &image_barrier.subresourceRange);
    test.End();
    test.Submit0Signal(test.cba);

    test.BeginB();
    // Uncommenting this will make the validation error go away:
    //image_barrier.srcStageMask = VK_PIPELINE_STAGE_2_TRANSFER_BIT;
    // Note that when the barrier is missing entirely, the CmdClearColorImage will not result in WRITE-AFTER-WRITE error and the test passes, even though CmdClearColorImage is also a write operation
    vk::CmdPipelineBarrier2KHR(test.h_cbb, &dep_info);
    vk::CmdClearColorImage(test.h_cbb, image, VK_IMAGE_LAYOUT_GENERAL, &m_clear_color, 1, &image_barrier.subresourceRange); // (optional)
    test.End();
    test.Submit0Wait(test.cbb, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT);

    test.DeviceWait();
}
```

**Comment by artem-lunarg** [comment]
@theakman2 Thanks for report. I'll check it. In general I'd recommend to open new issues for new use cases.
